src/Razor/src/Microsoft.AspNetCore.Razor.LanguageServer/Semantic/SemanticTokensCache.cs

```diff
@@ -0,0 +1,216 @@
+                    }
+
+                    tokenIndex += 5;
+                }
+
+                lineToTokensDict.Add(absoluteLine, lineTokens);
+            }
+
+            // Fill in the gaps for empty lines. When looking up tokens later, this helps us differentiate
+            // between empty lines vs. lines we don't have cached tokens for
+            for (var lineIndex = range.Start.Line; lineIndex < range.End.Line; lineIndex++)
+            {
+                if (!lineToTokensDict.TryGetValue(lineIndex, out _))
+                {
+                    lineToTokensDict.Add(lineIndex, Array.Empty<int>());
+                }
+            }
+        }
+
+        /// <summary>
+        /// Attempts to retrieve the cached tokens for a given document/semanticVersion/range. If the cache does
+        /// not have complete tokens for the range, it's possible to return only tokens associated with a portion
+        /// of the range.
+        /// </summary>
+        /// <param name="uri">The URI associated with the cached tokens.</param>
+        /// <param name="semanticVersion">The semantic version associated with the cached tokens.</param>
+        /// <param name="requestedRange">The requested range for the desired tokens.</param>
+        /// <param name="cachedTokens">If found, contains the cached range and cached tokens.</param>
+        /// <returns></returns>
+        internal bool TryGetCachedTokens(
+            DocumentUri uri,
+            VersionStamp semanticVersion,
+            Range requestedRange,
+            [NotNullWhen(true)] out (Range Range, List<int> Tokens)? cachedTokens)
+        {
+            if (!_cache.TryGetValue(uri, out var documentCache) ||
+                !documentCache.TryGetValue(semanticVersion, out var lineToTokensDict))
+            {
+                // No cached results found
+                cachedTokens = null;
+                return false;
+            }
+
+            var tokens = new List<int>();
+
+            // Keep track of where the cached range begins
+            Position? cachedRangeStart = null;
+
+            // Keep track of where the cached range ends
+            var cachedRangeNumLines = 0;
+
+            // Keeps track of the current empty line count (needed for calculating line offsets)
+            var emptyLineCount = 0;
+
+            // Retrieve the cached tokens associated with the passed-in range
+            for (var line = requestedRange.Start.Line; line < requestedRange.End.Line; line++)
+            {
+                if (!lineToTokensDict.TryGetValue(line, out var lineTokens))
+                {
+                    if (tokens.Count != 0)
+                    {
+                        // We already have some sort of cached results, but we've now reached
+                        // a gap in the cache which we have no tokens for.
+                        // Since we only want to return a continuous range, we'll exit early.
+                        break;
```

Comment:
same comment about extracting methods here.  the main probelm i have is keeping track of dataflow and jumps.  

note: you mention "exit early", but that's ambiguous.  it exits the loop, but then does more work.  "Exits early" sounds like it could just exit the method and do nothing.

Reply:
Good point - extracted a few methods and will update comment in my next push 👍